### PR TITLE
Allow other accounts to publish to SNS topics

### PR DESCRIPTION
**Pull request by jamieparkinson**
Required for https://github.com/wellcometrust/platform/issues/4163

### CODE REVIEW

File: sns/cross_account_topic_policy.tf [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
 
-data "aws_iam_policy_document" "cross_account_sns_topic_policy" {
-  policy_id = "__default_policy_ID"
+resource "aws_sns_topic_policy" "cross_account_topic_subscription_policy" {
+  count = "${local.has_cross_account_pubs || local.has_cross_account_subs ? 1 : 0}"
+  arn   = "${aws_sns_topic.topic.arn}"
+
+  policy = "${local.has_cross_account_subs && local.has_cross_account_subs
+    ? data.aws_iam_policy_document.cross_account_sns_topic_policy_pub_and_sub.json
+    : (local.has_cross_account_pubs
```
Comment (jamieparkinson) [inline sns/cross_account_topic_policy.tf]: This nonsense is to overcome the lack of conditional blocks and/or support for an empty list of Principals

Comment (alexwlchan) [reply sns/cross_account_topic_policy.tf]: Could there be a comment explaining it?

File: sns/cross_account_topic_policy.tf [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
+  has_cross_account_pubs = "${length(var.cross_account_publication_ids) > 0}"
 }
 
-data "aws_iam_policy_document" "cross_account_sns_topic_policy" {
-  policy_id = "__default_policy_ID"
+resource "aws_sns_topic_policy" "cross_account_topic_subscription_policy" {
+  count = "${local.has_cross_account_pubs || local.has_cross_account_subs ? 1 : 0}"
+  arn   = "${aws_sns_topic.topic.arn}"
+
+  policy = "${local.has_cross_account_subs && local.has_cross_account_subs
```
Comment (alexwlchan) [inline sns/cross_account_topic_policy.tf]: Why are you `&&`'ing a local with itself?

File: sns/cross_account_topic_policy.tf (hunk 1 above)
Comment (alexwlchan) [reply sns/cross_account_topic_policy.tf]: FWIW, we tend to break up this stuff with judicious use of locals.

**Review by alexwlchan — CHANGES_REQUESTED** [review]
Can we break up that monster conditional with some locals to make it clearer what it's doing?

File: sns/cross_account_topic_policy.tf (hunk 2 above)
Comment (jamieparkinson) [reply sns/cross_account_topic_policy.tf]: because bad at typing 🙃

File: sns/cross_account_topic_policy.tf [hunk 3]
```diff
@@ -1,14 +1,35 @@
 data "aws_caller_identity" "current" {}
 
-resource "aws_sns_topic_policy" "cross_account_topic_subscription_policy" {
-  count  = "${length(var.cross_account_subscription_ids) > 0 ? 1 : 0}"
+data "aws_region" "current" {}
+
+locals {
+  has_cross_account_subs = "${length(var.cross_account_subscription_ids) > 0}"
+  has_cross_account_pubs = "${length(var.cross_account_publication_ids) > 0}"
```
Comment (alexwlchan) [inline sns/cross_account_topic_policy.tf]: Let's increase the difference between these variables too:

```
locals {
  has_cross_account_subscriptions = ...
  has_cross_account_publications = ...
```

File: sns/cross_account_topic_policy.tf [hunk 4]
```diff
@@ -68,3 +93,66 @@ data "aws_iam_policy_document" "cross_account_sns_topic_policy" {
     sid = "ReportingAccess"
   }
 }
+
+data "aws_iam_policy_document" "cross_account_sns_topic_policy_publications" {
```
Comment (alexwlchan) [inline sns/cross_account_topic_policy.tf]: for consistency:

```suggestion
data "aws_iam_policy_document" "cross_account_sns_topic_policy_publications_only" {
```

File: sns/cross_account_topic_policy.tf [hunk 5]
```diff
@@ -44,6 +65,10 @@ data "aws_iam_policy_document" "cross_account_sns_topic_policy" {
 
     sid = "__default_statement_ID"
   }
+}
+
+data "aws_iam_policy_document" "cross_account_sns_topic_policy_subscriptions" {
```
Comment (alexwlchan) [inline sns/cross_account_topic_policy.tf]: for consistency:

```suggestion
data "aws_iam_policy_document" "cross_account_sns_topic_policy_subscriptions_only" {
```

**Review by alexwlchan — COMMENTED** [review]
A couple more suggestions for increased clarity.

File: RELEASE.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+RELEASE_TYPE: minor
+
+Add ability to specify accounts which can publish to an SNS topic
```
Comment (alexwlchan) [inline RELEASE.md]: How about:

> Allow AWS accounts to publish to an SNS topic in a different account. Use the `cross_account_publication_ids` parameter to the `sns` module.

**Review by alexwlchan — APPROVED** [review]
One suggestion on the changelog, otherwise looks good.

**Pull request merged**